File: DSA/Tree/Code/PreorderTraversal.java
// Preorder Traversal in Java

import java.util.*;

class Node {
    int data;
    Node left;
    Node right;
    
    Node(int data) {
        this.data = data;
        this.left = null;
        this.right = null;
    }
}

public class PreorderTraversal {
    
    public static void preorderRecursive(Node root) {
        if (root == null) return;
        
        System.out.print(root.data + " ");
        preorderRecursive(root.left);
        preorderRecursive(root.right);
    }
    
    public static void preorderIterative(Node root) {
        Stack<Node> stack = new Stack<>();
        stack.push(root);
        
        while (!stack.isEmpty()) {
            Node node = stack.pop();
            System.out.print(node.data + " ");
            
            if (node.right != null) {
                stack.push(node.right);
            }
            if (node.left != null) {
                stack.push(node.left);
            }
        }
    }
    
    public static void main(String[] args) {
        System.out.println("=== Preorder Traversal (Java) ===\n");
        
        // Create tree
        Node root = new Node(1);
        root.left = new Node(2);
        root.right = new Node(3);
        root.left.left = new Node(4);
        root.left.right = new Node(5);
        root.right.left = new Node(6);
        root.right.right = new Node(7);
        
        System.out.println("Tree structure:");
        System.out.println("         1");
        System.out.println("        / \\");
        System.out.println("       2   3");
        System.out.println("      / \\ / \\");
        System.out.println("     4  5 6  7\n");
        
        System.out.print("Recursive Preorder (Root-Left-Right): ");
        preorderRecursive(root);
        System.out.println();
        
        System.out.print("Iterative Preorder (using stack):     ");
        preorderIterative(root);
        System.out.println("\n");
        
        System.out.println("=== Preorder Characteristics ===");
        System.out.println("1. Root → Left subtree → Right subtree");
        System.out.println("2. Root is processed first");
        System.out.println("3. Useful for creating copy of tree");
        System.out.println("4. Useful for prefix expression evaluation\n");
        
        System.out.println("=== Complexity Analysis ===");
        System.out.println("Time Complexity:  O(n) - visit each node once");
        System.out.println("Space Complexity: O(h) - h = height (recursion/stack)");
        System.out.println("                   O(n) worst case (skewed tree)");
    }
}

File: DSA/Tree/Code/PreorderTraversal.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
    int data;
    struct Node* left;
    struct Node* right;
} Node;

Node* createNode(int data) {
    Node* node = (Node*)malloc(sizeof(Node));
    node->data = data;
    node->left = NULL;
    node->right = NULL;
    return node;
}

// Recursive preorder traversal: Root-Left-Right
void preorderRecursive(Node* root) {
    if (root == NULL) return;
    
    printf("%d ", root->data);
    preorderRecursive(root->left);
    preorderRecursive(root->right);
}

// Iterative preorder traversal using stack
typedef struct {
    Node* data[100];
    int top;
} Stack;

Stack* createStack() {
    Stack* stack = (Stack*)malloc(sizeof(Stack));
    stack->top = -1;
    return stack;
}

void push(Stack* stack, Node* node) {
    if (stack->top < 99) {
        stack->data[++stack->top] = node;
    }
}

Node* pop(Stack* stack) {
    if (stack->top >= 0) {
        return stack->data[stack->top--];
    }
    return NULL;
}

int isEmpty(Stack* stack) {
    return stack->top == -1;
}

void preorderIterative(Node* root) {
    Stack* stack = createStack();
    push(stack, root);
    
    while (!isEmpty(stack)) {
        Node* node = pop(stack);
        printf("%d ", node->data);
        
        if (node->right != NULL) {
            push(stack, node->right);
        }
        if (node->left != NULL) {
            push(stack, node->left);
        }
    }
}

int main() {
    printf("=== Preorder Traversal (C) ===\n\n");
    
    // Create tree
    Node* root = createNode(1);
    root->left = createNode(2);
    root->right = createNode(3);
    root->left->left = createNode(4);
    root->left->right = createNode(5);
    root->right->left = createNode(6);
    root->right->right = createNode(7);
    
    printf("Tree structure:\n");
    printf("         1\n");
    printf("        / \\\n");
    printf("       2   3\n");
    printf("      / \\ / \\\n");
    printf("     4  5 6  7\n\n");
    
    printf("Recursive Preorder (Root-Left-Right): ");
    preorderRecursive(root);
    printf("\n");
    
    printf("Iterative Preorder (using stack):     ");
    preorderIterative(root);
    printf("\n\n");
    
    printf("=== Preorder Characteristics ===\n");
    printf("1. Root → Left subtree → Right subtree\n");
    printf("2. Root is processed first\n");
    printf("3. Useful for creating copy of tree\n");
    printf("4. Useful for prefix expression evaluation\n\n");
    
    printf("=== Complexity Analysis ===\n");
    printf("Time Complexity:  O(n) - visit each node once\n");
    printf("Space Complexity: O(h) - h = height (recursion/stack)\n");
    printf("                   O(n) worst case (skewed tree)\n");
    
    return 0;
}
